Pre-release activities

kartikeyakirar opened the pull request:
part of https://github.com/insightsengineering/nestdevs-tasks/issues/48

- [x] Review and update:
      - README.md (check example code)
      - NEWS.md
- [x] Run urlchecker::url_check() to identify broken links and fix
- [x] Review functions:
       - @example tag, make sure it runs, fix if otherwise
       - Make sure functions has @return tag to document the return value
       - no \dontrun tag, replace with if(interactive()) if needed
- [x] Sanity check of all vignettes, make sure there is no typo, no wrong format, etc.
- [x] Run R CMD check --as-cran make sure everything pass
- [x] Package Title is not duplicated in Package Description in DESCRIPTION file (e.g. this happens in teal.slice currently)
- [x] All package names in Title and Description fields of DESCRIPTION file are quoted with '
- [x] You have checked the Package Release Template https://github.com/insightsengineering/teal.reporter/pull/205/files
- [x] Make sure there are no ::: in examples
- [x] Make sure all teal.* mentions are lower-cased and quoted
- [x] Make Sure inst/WORDLIST is minimalized
- [x] Make sure non-exported functions do not have examples
- [x] Make sure each link to our documentation hosted with pkgdown on github pages do not have /main/ in the address but it has /latest/ instead, so we always expose the documentation of the latest release and not what's currently on main branch but not yet released

chlebowa commented on R/FilterStates-utils.R:
add to -> can we keep the examples?

kartikeyakirar replied:
added [7feb212](https://github.com/insightsengineering/teal.slice/pull/506/commits/7feb212b81183ed335d45565112f9c0f9b1643ab)

chlebowa commented on R/FilterStatesDF.R:
add to -> can we keep the examples?

kartikeyakirar replied:
[7feb212](https://github.com/insightsengineering/teal.slice/pull/506/commits/7feb212b81183ed335d45565112f9c0f9b1643ab)

chlebowa commented on R/FilteredData-utils.R:
add to -> can we keep the examples?

kartikeyakirar replied:
Added

chlebowa replied:
:point_up_2: bump

chlebowa commented on R/FilteredData.R:
add to -> can we keep the examples?

chlebowa commented on R/FilteredDatasetDataframe.R:
add to -> can we keep the examples?

chlebowa commented on R/FilteredDatasetDefault.R:
add to -> can we keep the examples?

chlebowa commented on R/FilteredDatasetMAE.R:
add to -> can we keep the examples?

chlebowa commented on R/count_labels.R:
add to -> can we keep the examples?

chlebowa commented on R/utils.R:
add to -> can we keep the examples?
`check_simple_name` is fine without but the ones `check_ellipsis` and `make_c_call` are useful

kartikeyakirar replied:
[7feb212](https://github.com/insightsengineering/teal.slice/pull/506/commits/7feb212b81183ed335d45565112f9c0f9b1643ab)

chlebowa commented on R/variable_types.R:
```diff
@@ -6,29 +6,7 @@
 #'
 #' @return (atomic vector of `character`) classes of `columns` from provided `data`
```
```suggestion
#' @return Character vector of classes of `columns` from provided `data`.
```

Also change argument descriptions:
```
#' @param data (`matrix` or `data.frame`-like) object to determine variable types from.
#' @param columns (`character`) vector of columns in `data` to get classes from.
#'   Set to `NULL` to get classes of all columns.
```

kartikeyakirar replied:
updated here [781e30b](https://github.com/insightsengineering/teal.slice/pull/506/commits/781e30b7752b481640589d25979370c1f12fe74f)

chlebowa commented on R/variable_types.R:
I am confused by this function. It seems that most methods are redundant and there is no handling of unsupported classes. I am also unfamiliar with the `DataTable` class, at least in the context of this package.

I will open a new issue for this. In the mean time, please adjust the documentation a little.

chlebowa commented on README.md:
```diff
@@ -53,7 +58,7 @@ To understand how to use this package, please refer to the [Introduction to teal
 ## Getting help
 
-If you encounter a bug or you have a feature request - please file an issue. For questions, discussions and staying up to date, please use the "teal" channel in the [`pharmaverse` slack workspace](https://pharmaverse.slack.com).
+If you encounter any bug or have a feature request, please file an issue. For questions, discussions, and updates, use the `teal` channel in the workspace.[`pharmaverse` slack workspace](https://pharmaverse.slack.com).
```
```suggestion
If you encounter a bug or have a feature request, please file an issue. For questions, discussions, and updates, use the `teal` channel in the [`pharmaverse` slack workspace](https://pharmaverse.slack.com).
```

chlebowa commented on _pkgdown.yml:
```diff
@@ -21,8 +21,8 @@ navbar:
 reference:
-  - title: Teal filter-panel API
-    desc: Functions used initialize filter-panel and to modify its states.
+  - title: "teal filter-panel API"
```
Why remove the backticks?

kartikeyakirar replied:
I have added backticks. Earlier I was getting some syntax issues on cran check but its not there anymore.

chlebowa commented on NEWS.md:
```diff
@@ -59,7 +59,7 @@
 ### Bug fixes
 
 * Fixed an error where the `RangeFilterState` produced an error when using `bootstrap 4`.
-* Fixed a bug that caused the range slider to omit values selected programmatically through the filter API.
```
Is the word "programmatically" removed only to trim the wordlist or is there another reason?

chlebowa commented on vignettes/filter-panel-for-developers.Rmd:
```diff
@@ -51,12 +51,12 @@ Each `teal_slice` object contains all the information necessary to:
    - `multiple` (only for `ChoiceFilterState`) - allows multiple values to be selected
    - `fixed` - forbids changing state of the `FilterState`
-   - `anchored` - forbids removal of the `FilterState`
+   - `anchored` - forbids the removal of the `FilterState`
```
```suggestion
   - `anchored` - forbids removing the `FilterState`
```

chlebowa commented on vignettes/filter-panel-for-developers.Rmd:
```diff
@@ -12,10 +12,10 @@ vignette: >
 These methods include:
 
-- `get_filter_state` - returns the current state of filters in form of `teal_slices` object
+- `get_filter_state` - returns the current state of filters in the form of `teal_slices` object
```
```suggestion
- `get_filter_state` - returns current state of filters as `teal_slices` object
```

chlebowa commented on vignettes/filter-panel-for-developers.Rmd:
```diff
@@ -12,10 +12,10 @@ vignette: >
 
-- `get_filter_state` - returns the current state of filters in form of `teal_slices` object
+- `get_filter_state` - returns the current state of filters in the form of `teal_slices` object
 - `set_filter_state` - adds or modifies the filters based on `teal_slices` object
```
```suggestion
- `set_filter_state` - adds or modifies filters based on `teal_slices` object
```

chlebowa commented on vignettes/filter-panel-for-developers.Rmd:
```diff
@@ -51,12 +51,12 @@ Each `teal_slice` object contains all the information necessary to:
    - `choices` - determines the set of values or the range that can be selected from
```
```suggestion
   - `choices` - determines set of values or range that can be selected from
```

chlebowa commented on vignettes/filter-panel-for-developers.Rmd:
```diff
@@ -51,12 +51,12 @@ Each `teal_slice` object contains all the information necessary to:
 In addition, every `teal_slice` object has an `id`.
 
-It is impossible to create `FilteredData` with slices with duplicated `id`s.
+It is impossible to create `FilteredData` with slices with  duplicates `id`s.
```
```suggestion
It is impossible to create `FilteredData` with slices with duplicated `id`s.
```

chlebowa commented on vignettes/filter-panel-for-developers.Rmd:
```diff
@@ -138,7 +138,7 @@ set_filter_state(
 
 2. `exclude_varnames` defines which columns in the used data sets are **not** allowed to be filtered on.
-In following example all variables except the four will be available to choose from.
+In the following example, all variables except the four will be available to choose from.
```
```suggestion
In the following example all variables except the four will be available to choose from.
```

chlebowa commented on vignettes/filter-panel-for-developers.Rmd:
````diff
@@ -151,7 +151,7 @@ set_filter_state(
 ```
 
-3. `count_type` defines how observations counts displayed in filter cards
+3. `count_type` defines how observation counts displayed in filter cards
````
```suggestion
3. `count_type` defines how observation counts are displayed in filter cards
```

chlebowa commented on vignettes/teal-slice-classes.Rmd:
```diff
@@ -117,9 +117,9 @@ The following `FilteredDataset` child classes are currently implemented:
 `FilterStates` serves two `shiny`-related purposes:
 
-- `ui/srv_add` allows adding a `FilterState` for a selected variable. The variables included in the module are the filterable column names of the provided data set. Selecting a variable adds a `FilterState` to the `reactiveVal` stored in the `private$state_list` private field. The subtype of the created `FilterState` automatically determined based on the class of the selected column.
+- `ui/srv_add` allows adding a `FilterState` for a selected variable. The variables included in the module are the filterable column names of the provided data set. Selecting a variable adds a `FilterState` to the `reactiveVal` stored in the `private$state_list` private field. The specific variation of the created `FilterState` is automatically determined based on the class of the selected column ...
```
What's wrong with "subtype"?

chlebowa replied:
I believe "Make Sure inst/WORDLIST is minimalized" was intended to mean make sure the wordlist does not contain only words that actually occur in the  package. "Subtype" is a valid term, as are supertype, subclass, and superclass.

chlebowa commented on vignettes/teal-slice-classes.Rmd:
```diff
@@ -136,18 +136,18 @@ A `FilterState` is created as follows:
 All child classes handle missing values, and `RangedFilterState` also handles infinite values.
 
-The `FilterState` constructor also takes the `extract_type` argument, which determines how the call is constructed `extract_type` can be unspecified, `"matrix"` or `"list"` and its value corresponds to the type of the variable prefix in the returned condition call. If `extract_type` is `"list"`, the variable in the condition call is `<dataname>$<varname>`, while for `"matrix"` it would be `<datan ...
+The `FilterState` constructor also takes the `extract_type` argument, which determines how the call is constructed `extract_type` can be unspecified, `"matrix"` or `"list"` and its value corresponds to the type of the variable prefix in the returned condition call. If `extract_type` is `"list"`, the variable in the condition called is `<dataname>$<varname>`, while for `"matrix"` it would be `<dat ...
```
```suggestion
The `FilterState` constructor also takes the `extract_type` argument, which determines how the call is constructed `extract_type` can be unspecified, `"matrix"` or `"list"` and its value corresponds to the type of the variable prefix in the returned condition call. If `extract_type` is `"list"`, the variable in the condition call is `<dataname>$<varname>`, while for `"matrix"` it would be `<dataname>[, "<varname>"]`.
```

kartikeyakirar commented on R/FilterState-utils.R:
I completely agree these solutions are helpful in understanding the functionality. Here are some suggestions to include examples for internal functions in the package without  causing issues:

**Conditional Execution:** Use \dontrun{} or \donttest{} tags in your documentation to prevent certain examples from running during CRAN checks,  as it was asked to remove these tags I am not sure if its due cran or if we don't prefer keeping it in pkg.

**Vignettes:** Create seperate vignettes for advanced guidance, particularly useful for developers, illustrating the use of internal functions. keeping eval = FALSE https://r-pkgs.org/vignettes.html#:~:text=CRAN's%20incoming%20and%20ongoing%20checks,doomed%20to%20fail%20on%20CRAN.

**External Documentation:** Consider hosting detailed examples on an external repo/website as a last resort for advanced users.

**Unit Tests**: Include examples in unit tests, adding detailed comments for clarity.

chlebowa replied:
> **Conditional Execution:** Use \dontrun{} or \donttest{} tags in your documentation to prevent certain examples from running during CRAN checks, as it was asked to remove these tags I am not sure if its due cran or if we don't prefer keeping it in pkg.

CRAN now disregards those tags.

> **External Documentation:** Consider hosting detailed examples on an external repo/website as a last resort for advanced users.

That will never be properly maintained. If we keep it, it should be within the package.

> **Unit Tests**: Include examples in unit tests, adding detailed comments for clarity.

I doubt anyone would look there.

> **Vignettes:** Create seperate vignettes for advanced guidance, particularly useful for developers, illustrating the use of internal functions. keeping eval = FALSE

I like this idea. Please include a note in the docs of the relevant files that points to this vignette.

kartikeyakirar replied:
@chlebowa I have added vignette here [34fe8b9](https://github.com/insightsengineering/teal.slice/pull/506/commits/34fe8b9b5499eec98bc680224732c1d7e24ac3bc) let me know your thought on this.

pawelru commented on .github/ISSUE_TEMPLATE/cran-release.yaml:
```diff
@@ -0,0 +1,124 @@
+assignees:
+  - KlaudiaBB
+  - cicdguy
+  - shajoezhu
```
I don't think we need to call @shajoezhu for release of this repo

donyunardi replied:
Did you mean me instead of @shajoezhu ?

Also, this .yaml file is propagated from either insightsengineering/.github or insightsengineering/.github-private. If we make changes here, they will still be overwritten during the next propagation. If modifications are necessary, they should be made in those repos.

donyunardi replied:
Similar to https://github.com/insightsengineering/teal.data/pull/240/files#r1442001265, let's leave it for now.

Related to https://github.com/insightsengineering/idr-tasks/issues/722

chlebowa commented on NEWS.md:
```diff
@@ -59,7 +59,7 @@
 
 * Fixed an error where the `RangeFilterState` produced an error when using `bootstrap 4`.
-* Fixed a bug that caused the range slider to omit values selected programmatically through the filter API.
+* Fixed a bug that caused the range slider to omit values selected via filter API.
```
```suggestion
* Fixed a bug that caused the range slider to omit values selected by the filter API.
```

chlebowa commented on vignettes/internal_functions_example.Rmd:
1. Separate examples into sections: `FilteredData, `FilteredDataset`, `FilterState`, other.
2. Remove "Example for" from section titles.
3. Rename file to `internal_function_examples.Rmd.
4. Change title to _Examples for internal functions_.

kartikeyakirar replied:
> Separate examples into sections: FilteredData, FilteredDataset, FilterState`, other.

I originally placed these under the suggested section, but I removed them for better navigation through the indexes.
<img width="350" alt="image" src="https://github.com/insightsengineering/teal.slice/assets/6700955/5bb56b2a-b135-4611-99af-20eeb253b1df">

kartikeyakirar replied:
file is too large to navigate via section

chlebowa replied:
What do you mean?

kartikeyakirar replied:
Sorry, I created the section incorrectly and as a result, the examples title didn't show up on the index. I will add the section again.

kartikeyakirar replied:
updated Rmd
<img width="346" alt="image" src="https://github.com/insightsengineering/teal.slice/assets/6700955/af41d697-1ac2-41b7-acc5-6d1a23933a03">

chlebowa replied:
OK, so what's wrong with this?

kartikeyakirar replied:
Nothing, the screenshot is an updated version, I pushed changes for it .

chlebowa replied:
> file is too large to navigate via section

So this is no longer relevant?

kartikeyakirar replied:
yes

chlebowa commented on vignettes/internal_functions_example.Rmd:
```diff
@@ -0,0 +1,776 @@
+---
+
+## Introduction
+This vignette designed to showcase examples for internal classes, methods, and functions of  teal.slice, specifically designed developers.
```
```suggestion
This vignette contains examples for internal classes, methods, and functions of  `teal.slice`. 
For developer use only.
```

chlebowa commented on R/FilterState-utils.R:
```diff
@@ -19,37 +19,10 @@
-#'   ),
-#'   extract_type = "matrix"
-#' )
+#' @seealso check examples here `vignette("internal_functions_example", package = "teal.slice")`.
```
```suggestion
#' @seealso examples found here: `vignette("internal_functions_example", package = "teal.slice")`.
```

chlebowa reviewed (changes requested):
The vignette with examples is incomplete and multiple comments have not been addressed.

kartikeyakirar commented:
> The vignette with examples is incomplete and multiple comments have not been addressed.

Apologies for missing some open comments. it got hidden and I thought I addressed them 😓.
I have updated PR.

chlebowa reviewed (changes requested):
Please arrange sections in examples vignette according to class hierarchy.

Make sure help pages of _all_ functions that have examples moved to vignette point to the vignette.

chlebowa commented on R/FilteredDataset-utils.R:
You missed some examples here.

kartikeyakirar replied:
I moved the example to the bottom of the documentation and removed the `:::` as `init_filtered_dataset` is an exported function.

chlebowa replied:
Indeed, my mistake.

chlebowa commented on R/calls_combine_by.R:
missing `@seealso`

kartikeyakirar replied:
mistakenly, I added a `####` (heading 4) for the function `calls_combine_by` in the vignette. As a result, it was not visible in the index. and I missed this entry, I have added @seealso and updated the vignette as well. Thank you.

chlebowa commented on R/teal_slices.R:
```diff
@@ -271,8 +271,9 @@ setdiff_teal_slices <- function(x, y) {
   checkmate::assert_list(x)
-  if (all(vapply(x, is.atomic, logical(1L)))) {
-    return(Filter(Negate(is.null), x)[[1L]])
+  xnn <- Filter(Negate(is.null), x)
+  if (all(vapply(xnn, is.atomic, logical(1L)))) {
+    return(xnn[[1L]])
```
This is already in on `main`, I worry there are some merging artefacts on this branch.

kartikeyakirar replied:
branch is up to date with main. let try with changing base branch.

kartikeyakirar replied:
its not there anymore.

chlebowa commented on R/FilteredDataset-utils.R:
Internal function examples remain.

chlebowa reviewed (dismissed):
Everything seems to check out.
Blocking until we resolve two more issues: 
- the discussion about #510 
- update images in the Filter Panel Classes vignette: they still use `DefaultFilteredDataset` where they should use `DataframeFilteredDataset`, an oversight from a previous PR. @gogonzo do you still have templaets for those images or should I recreate them?

pawelru commented:
> @gogonzo do you still have templaets for those images or should I recreate them?

I do recall myself asking to commit it so as to make it available for the situation like this one. Isn't it this one? https://github.com/insightsengineering/teal.slice/blob/main/inst/UMLs/filter_panel.xml It probably needs to be open with diagrammer app

chlebowa commented:
> Isn't it this one?

Yes! Thank you, I will apply the changes there.

chlebowa commented:
No, unfortunately this file seems outdated.

gogonzo commented:
> * DataframeFilteredDataset

https://github.com/insightsengineering/teal.slice/pull/513/files

donyunardi commented:
> * the discussion about [use demos to store examples for internal functions #510](https://github.com/insightsengineering/teal.slice/pull/510)

Bringing in the discussion from Slack:
* remove the internal-function-example vignette.
* we'll use `getFromNamespace()` to substitute `:::` and keep the examples close to the function. 
* Once we have done this, create a parallel branch to apply the `demo` feature, just in case. We create the branch after the `getFromNamespace()` implementation so that there won't be any conflict when merging.

kartikeyakirar commented:
> Should` we skip MAE tests if MultiAssayExperiment is not installed? 

+1, I think we should also, do the same for the `SummarizedExperiment`  package examples.

chlebowa commented:
That's not a reason to skip tests. `MAE` is in `Suggests` so installing is the responsibility of the person who wants to run tests, i.e. not the end user. If someone uses `teal` with `MAE` data, they most likely have the package installed.

averissimo commented:
@chlebowa According to CRAN's [Writing R Extensions](https://cran.r-project.org/doc/manuals/r-release/R-exts.html#Suggested-packages) it is _(at least that's my interpretation)_. But that section it's also a bit contradictory.

That being said, this is most likely not a problem as CRAN's check system has access to Bioc packages, and current NEST packages such as {rtables} have packages in Suggests that are used in tests without any conditional check.

We're just not strictly following guidelines.

![image](https://github.com/insightsengineering/teal.slice/assets/211358/a11a6f05-6022-4fd8-a69d-cde2681578ff)

pawelru commented:
Skip if no (suggested) package installed will be required after: https://github.com/insightsengineering/idr-tasks/issues/713. Please see the reasoning there

averissimo commented:
Given @pawelru comment I've pushed a branch with the pre-emptive work I did to the repo on skipping tests/examples for the team to continue to discuss.

The changes can be checked [here](https://github.com/insightsengineering/teal.slice/compare/pre-release-cleanup%40main...soft-dependencies%40pre-release-cleanup%40main?diff=split&w=) (`soft-dependencies@pre-release-cleanup@main` branch).

They pass the checks both in CI (normal workflow) and locally (without {MAE} and {SE}).

pawelru commented:
@averissimo would you mind opening a (draft) PR to collect comments there separately?

pawelru commented:
I have checked the output of R CMD CHECK GHAction for possible notes that have not been blocked and I found one:
```
* checking Rd cross-references ... NOTE
Unknown package ‘teal’ in Rd xrefs
```

Is this a problem? If yes - can we fix it somehow (and also ask IDR to add it to the block-list and propagate).

(I do encourage to do the same thing for the rest of clean-ups)

kartikeyakirar commented:
```
* checking Rd cross-references ... NOTE
Unknown package ‘teal’ in Rd xrefs
```
In the `teal_slices.R` documentation, there is a note referencing the` teal `package, specifically linking to` [teal::slices_store] and [teal::slices_restore]. `. Thereis check [` _R_CHECK_RD_XREFS_`](https://cloud.r-project.org/doc/manuals/r-patched/R-ints.html#index-_005fR_005fCHECK_005fRD_005fXREFS_005f). If set to true, it verifies cross-references in .Rd files against CRAN packages, with the default being true.

I believe it shouldn't pose any issues, but you can consider setting R_CHECK_RD_XREFS to FALSE.

chlebowa commented on R/FilterState.R:
```diff
@@ -2,13 +2,13 @@
 #'
 #'
-#' @title `FilterState` Abstract Class
+#' @title `FilterState` abstract Class
```
I believe it's redundant.

+1 for _somebody_ going through all files and applying:
- good title (no tag)
- good description (no tag)
- proper details if description unreasonably long (no tag) (more than one simple sentence is too much)

Note that in `teal.code`, `teal.data` and `teal` we have man files built from several `R` files. In those files `@name` and `@rdname` are necessary and `@details` is quite helpful and should not be removed.

kartikeyakirar replied:
`teal.code` is already on cran and  there is one instance where `@description` tag is used. I will revisit the `teal.data` and raise a separate PR  for removing the tag (as pre-release [PR](https://github.com/insightsengineering/teal.data/pull/240) is merged to main).   And  for` teal` and `teal.slice ` I will include changes in existing pre-release activity PR .

chlebowa replied:
> Note that in `teal.code`, `teal.data` and `teal` we have man files built from several `R` files. In those files `@name` and `@rdname` are necessary and `@details` is quite helpful and should not be removed.

Like I said
> Note that in teal.code, teal.data and teal we have man files built from several R files. In those files @name and @rdname are necessary and @details is quite helpful and should not be removed.

m7pr replied:
@kartikeyakirar is the `@title` finally removed in this PR for this file?

pawelru commented on R/FilteredDatasetMAE.R:
Please correct `state` param description of `set_filter_state` method - `list` type of data seems to be invalid as per assert inside the body of it

kartikeyakirar replied:
Thank you for catching that. I will also check in the other class as well.

kartikeyakirar replied:
here [a8e59aa](https://github.com/insightsengineering/teal.slice/pull/506/commits/a8e59aa6a6eee91b2b8864b00da98f483e80401e)

pawelru commented on R/FilterStates.R:
(same as above)
Please correct `state` param description of `set_filter_state` method - list type of data seems to be invalid as per assert inside the body of it

kartikeyakirar replied:
updated here [a8e59aa](https://github.com/insightsengineering/teal.slice/pull/506/commits/a8e59aa6a6eee91b2b8864b00da98f483e80401e)

pawelru commented on .lintr:
```diff
@@ -1,5 +1,6 @@
   cyclocomp_linter = NULL,
-  object_usage_linter = NULL
+  object_usage_linter = NULL,
+  indentation_linter = NULL
```
why disable this?

kartikeyakirar replied:
The code was restyled automatically by a GitHub auto-action when code is changed in accordance with indentation warnings by `linter`. This issue of indentation warnings was encountered across most of our packages, that's why  `indentation_linter ` is set to `NULL.`

warnings:
<img width="535" alt="image" src="https://github.com/insightsengineering/teal.slice/assets/6700955/97f545f5-c8f6-4b97-83d6-75d75ba2dd3a">

averissimo replied:
These `indentation_linter` are most likely due to a conflict on lint and styling on `if` clauses

I've been using the (accepted) style below to get around the conflict. I found this issue a while ago in lintr repo (with a known face commenting there :smile:) https://github.com/r-lib/lintr/issues/2034

I can make the changes on a PR targetting this branch to solve this and avoid having the rule disabled

````r
     # lint error
      } else if (setequal(private$get_selected(), private$get_choices()) &&
        !anyNA(private$get_selected(), private$get_choices())) {
        TRUE
      } 
     # No lint error and styler doesn't reframe this
      } else if (
        setequal(private$get_selected(), private$get_choices()) &&
          !anyNA(private$get_selected(), private$get_choices())
      ) {
        TRUE
      }
````

kartikeyakirar replied:
Thank you, @averissimo, for posting it here. that would be grate if the code can be formatted in a way that both {lintr} and {styler} accept, and looking forward to your PR for this. In my opinion, this should be handled as a separate issue since it affects all packages.  so, we can keep track of all the repositories  changes . What do you think?

>  I think it would be nice to report it somewhere (probably to lintr folks as a false positive) adding a minimal reprex based on the cases reported. I think this is the most common setup that should work.

@pawelru   should we duplicate https://github.com/r-lib/lintr/issues/2034 issue to keep track of it in `Coredev`

averissimo replied:
I did a quick PR that targets this branch, but it can easily target `main` instead

pawelru replied:
> should we duplicate [r-lib/lintr#2034](https://github.com/r-lib/lintr/issues/2034) issue to keep track of it in `Coredev`

IDK. Probably not. But definitely worth to add this as a code-comment what's the reason behind a given exclude

averissimo commented:
We could import the documentation from `teal.data:::topological_sort` to avoid having duplicate titles / description / params

`@examples` are necessary as that tag is not copied with `@inherit`

WDYT?

note: It uses double colon (despite being a non-exported function) and it seems to correctly import the documentation

```diff
diff --git a/R/FilteredData-utils.R b/R/FilteredData-utils.R
index a6b0cd7f..40a6cb1f 100644
--- a/R/FilteredData-utils.R
+++ b/R/FilteredData-utils.R
@@ -180,14 +180,7 @@ toggle_title <- function(input_id, titles, one_way = FALSE) {
   invisible(NULL)
 }
 
-#' Topological graph sort
-#'
-#' Graph is a list which for each node contains a vector of child nodes
-#' in the returned list, parents appear before their children.
-#'
-#' Implementation of `Kahn` algorithm with a modification to maintain the order of input elements.
-#'
-#' @param graph (named `list`) list with node vector elements
+#' @inherit teal.data::topological_sort description details params title
 #' @examples
 #' # use non-exported function from teal.slice
 #' topological_sort <- getFromNamespace("topological_sort", "teal.slice")
```

kartikeyakirar commented on R/FilteredData-utils.R:
```diff
@@ -188,55 +188,14 @@ toggle_title <- function(input_id, titles, one_way = FALSE) {
-#' teal.slice:::topological_sort(list(A = c(), B = c("A"), C = c("B"), D = c("A")))
-#' teal.slice:::topological_sort(list(D = c("A"), A = c(), B = c("A"), C = c("B")))
-#' teal.slice:::topological_sort(list(D = c("A"), B = c("A"), C = c("B"), A = c()))
+#' # use non-exported function from teal.slice
```
I think it's much better from a maintenance point of view. The only downside is that it's not self-contained. (https://r-pkgs.org/man.html#inheriting-arguments). but I don't  it as issue function definition itself is imported form `teal.data`.

updated changes here [aa6e774](https://github.com/insightsengineering/teal.slice/pull/506/commits/aa6e774b875e5da73a3c477202a2bd37b149d1b0)

averissimo replied:
Great! Hoping that this doesn't trigger any complaint from CRAN as you said, this is much better for maintenance

chlebowa commented on R/FilteredDataset-utils.R:
```diff
@@ -43,7 +43,7 @@
 #' # MAEFilteredDataset example
 #' library(MultiAssayExperiment)
-#' data(miniACC)
+#' data(miniACC, package = "MultiAssayExperiment")
```
Isn't `package` only necessary if the package isn't attached?

averissimo replied:
Yes, this was a strange thing, I just noticed that there's a typo and reverted this change.

My environment didn't have `MultiAssayExperiment` so the typo was not detected before the merge.

averissimo replied:
Corrected here: 511ede211560e99bfdac3eb0637f28cc1c74310b

pawelru commented on R/FilterPanelAPI.R:
```diff
@@ -1,20 +1,24 @@
+#' filter panel API.
 #'
 #' @examples
 #' library(teal.slice)
```
```suggestion
```

pawelru commented on R/FilteredData.R:
```diff
@@ -52,13 +57,43 @@
+#'
+#' utils::data(miniACC, package = "MultiAssayExperiment")
+#' datasets <- FilteredData$new(list(iris = iris, mae = miniACC))
+#' fs <-
+#'   teal_slices(
+#'     teal_slice(
+#'       dataname = "iris", varname = "Sepal.Length", selected = c(5.1, 6.4),
+#'       keep_na = TRUE, keep_inf = FALSE
+#'     ),
+#'     teal_slice(
+#'       dataname = "iris", varname = "Species", selected = c("setosa", "versicolor"),
+#'       keep_na = FALSE
+#'     ),
+#'     teal_slice(
+#'       dataname = "mae", varname = "years_to_birth", selected = c(30, 50),
+#'       keep_na = TRUE, keep_inf = FALSE
+#'     ),
+#'     teal_slice(dataname = "mae", varname = "vital_status", selected = "1", keep_na = FALSE),
+#'     teal_slice(dataname = "mae", varname = "gender", selected = "female", keep_na = TRUE),
+#'     teal_slice(
+#'       dataname = "mae", varname = "ARRAY_TYPE",
+#'       selected = "", keep_na = TRUE, datalabel = "RPPAArray", arg = "subset"
+#'     )
+#'   )
```
```suggestion
#' fs <- teal_slices(
#'   teal_slice(
#'     dataname = "iris", varname = "Sepal.Length", selected = c(5.1, 6.4),
#'     keep_na = TRUE, keep_inf = FALSE
#'   ),
#'   teal_slice(
#'     dataname = "iris", varname = "Species", selected = c("setosa", "versicolor"),
#'     keep_na = FALSE
#'   ),
#'   teal_slice(
#'     dataname = "mae", varname = "years_to_birth", selected = c(30, 50),
#'     keep_na = TRUE, keep_inf = FALSE
#'   ),
#'   teal_slice(dataname = "mae", varname = "vital_status", selected = "1", keep_na = FALSE),
#'   teal_slice(dataname = "mae", varname = "gender", selected = "female", keep_na = TRUE),
#'   teal_slice(
#'     dataname = "mae", varname = "ARRAY_TYPE",
#'     selected = "", keep_na = TRUE, datalabel = "RPPAArray", arg = "subset"
#'   )
#' )
```

pawelru commented on R/FilteredData.R:
```diff
@@ -52,13 +57,43 @@
+#'     ),
+#'     teal_slice(dataname = "mae", varname = "vital_status", selected = "1", keep_na = FALSE),
+#'     teal_slice(dataname = "mae", varname = "gender", selected = "female", keep_na = TRUE),
+#'     teal_slice(
+#'       dataname = "mae", varname = "ARRAY_TYPE",
+#'       selected = "", keep_na = TRUE, datalabel = "RPPAArray", arg = "subset"
+#'     )
```
this gave me a warning
```
filters for columns: ARRAY_TYPE excluded from mae
```

is this intended? Can we make it a warning-fee example?

pawelru replied:
Same for example in `MAEFilteredDataset.Rd`

chlebowa replied:
This is a reoccurring problem with how `excluded_variables` (or whatever the name) is set.
The only person who can fix this quickly is @gogonzo.

m7pr replied:
Started a separate issue for that https://github.com/insightsengineering/teal.slice/issues/531

m7pr commented on R/FilterState.R:
```diff
@@ -20,22 +21,21 @@
 #' `selected`, `varname`, `dataname`, `labels`, `na_count`, `keep_na` and other
 #' variable type specific fields (`keep_inf`, `inf_count`, `timezone`).
-#' Object contains also shiny module (`ui` and `server`) which manages
+#' Object also contains a shiny module (UI and server) which manages the
```
```suggestion
#' Object also contains a `shiny` module (UI and server) which manages the
```

@kartikeyakirar can you do GLOBAL FIND for `shiny` and maybe other packages (listed in DESCRIPTION) and make sure they usage of each packages in the documentation is quoted with `\code{}` (or backticks) or `\pkg{}`?

kartikeyakirar replied:
Thanks, I have checked all NEST packages, but I will double-check to ensure nothing was missed.

m7pr replied:
yes, just check for packages listed in DESCRIPTION

m7pr commented on R/FilterState.R:
```diff
@@ -46,26 +46,24 @@ FilterState <- R6::R6Class( # nolint
     #' @param ... additional arguments to be saved as a list in `private$extras` field
     #'
-    #' @return self invisibly
+    #' @return `self` invisibly
```
@kartikeyakirar does `self` needs to be quoted with backticks? Like it was an object. What if you rewrite to something like `Object of class \code{FilterState} (invisibly).`

chlebowa replied:
:exclamation: `self` is not the same as object of class `FilterState`.

I have no opinion on quoting `self` ATM.

m7pr replied:
I don't think `self` invisibly is enough in here. `Object of class \code{FilterState} (invisibly).` or at least `\code{FilterState} (invisibly)` to highlight the class of the returned value.

chlebowa replied:
`self` is more specific because it means _this very instance_ is returned, not just any `FilterState`. Needless to say, `self` implies the class rather unambiguously.

kartikeyakirar replied:
In the context of R6 class documentation, the term self refers to the object itself, and it's common to use backticks to indicate that it's a specific term or object in the code. In r6 documentation also its quoted with backticks https://github.com/search?q=repo%3Ar-lib%2FR6%20self&type=code

m7pr replied:
ok team, you convinced me my suggestion is not needed in here. resolving this one

m7pr commented on R/FilterState.R:
```diff
@@ -141,7 +139,7 @@ FilterState <- R6::R6Class( # nolint
     #' - `anchored` state is prevented from removing state
     #'
-    #' @param state a `teal_slice` object
+    #' @param state (`teal_slice`) object
```
@kartikeyakirar in `FilterState` where you have `@param slice`  we just write `(teal_slice)` without `object`. Can you unify that? I would opt for the removal of `object` in `@param` specifications

kartikeyakirar replied:
done [95ced22](https://github.com/insightsengineering/teal.slice/pull/506/commits/95ced22df80e0414a42be5ca506cdcd43a3e9a92)

m7pr commented on R/FilterStateChoices.R:
```diff
@@ -1,16 +1,24 @@
 #'   x = c(LETTERS, NA),
 #'   slice = teal_slice(varname = "x", dataname = "data")
 #' )
-#' shiny::isolate(filter_state$get_call())
```
@pawelru I've seen you wanted to move `shiny` from Depends to Imports in `teal.slice` https://github.com/insightsengineering/teal.slice/issues/529
If we decide to do such a change, we should either put `shiny::` prefix in examples or add `library(shiny)` at the beginning of examples. Anyway I would opt for leaving `shiny::` prefix in examples due to this expected transition

m7pr replied:
CC @kartikeyakirar

chlebowa replied:
Not required, we still have `@import shiny`.

m7pr replied:
I though `@import shiny` is a `roxygen2` tag that will add a `import(shiny)` entry to NAMESPACE. This allows functions of teal.slice to be able to use functions from shiny package, but it does not allow global namespace to be able to use shiny functions. 

When you load `teal.slice`
```r
> library(teal.slice)
Loading required package: shiny
Warning message:
package ‘shiny’ was built under R version 4.3.1 
```
Shiny is loaded because it's in `Depends` field of the description, not because we have `@import shiny` roxygen2 tag.

We also have `import(R6)` but `R6` is only in `Imports` hence you can not directly call R6 functions

```r
> Queue <- R6Class("Queue",
+                  public = list(
+                    initialize = function(...) {
+                      for (item in list(...)) {
+                        self$add(item)
+                      }
+                    },
+                    add = function(x) {
+                      private$queue <- c(private$queue, list(x))
+                      invisible(self)
+                    },
+                    remove = function() {
+                      if (private$length() == 0) return(NULL)
+                      # Can use private$queue for explicit access
+                      head <- private$queue[[1]]
+                      private$queue <- private$queue[-1]
+                      head
+                    }
+                  ),
+                  private = list(
+                    queue = list(),
+                    length = function() base::length(private$queue)
+                  )
+ )
Error in R6Class("Queue", public = list(initialize = function(...) { : 
  could not find function "R6Class"
```

Even though examples may pass during the build without `library(shiny)` if it's only in Imports, I think when user runs examples in his/her section this will fail, THE MOMENT WE move shiny from Depends to Imports.

chlebowa replied:
> global namespace

What is that? :raised_eyebrow:

m7pr replied:
A mind short cut of global `search()` path

chlebowa replied:
> Shiny is loaded because it's in `Depends` field of the description, not because we have `@import shiny` roxygen2 tag.

Wrong. It is _attached_ because it's in `Depends`, the default message is somewhat misleading. All it takes to load a namespace is to use `shiny::isolate` once.

We have `@import shiny` on top of `Depends: shiny` because that is demanded by checks: all packages in `Depends` must be imported from.

m7pr replied:
I think we have `Depends: shiny` so that we do not need to put `library(shiny)` in every example that uses any `shiny` function (like `ShinyApp`). If we remove `shiny` from `Depends` you will need to bring back either `library(shiny)` or prefix every `shiny` function call with `shiny::`.

Running `shiny::isolate()` once does not allow to run other functions from `shiny` package without `library()` or prefixing.

```r
> shiny::isolate()
Error in contextFunc() : argument "expr" is missing, with no default
> shinyApp
Error: object 'shinyApp' not found
```

chlebowa replied:
> I think we have `Depends: shiny` so that we do not need to put `library(shiny)` in every example

You are mistaken. It is because `shiny` contains the core functionalities that `teal` builds upon.

chlebowa replied:
> Running `shiny::isolate()` once does not allow to run other functions from `shiny` package without `library()` or prefixing.

This is exactly the difference between loading and attaching.

pawelru replied:
> P.S. IMO, all packages except teal could have shiny in imports.

In general I agree with this statement with exception of `teal` and also `teal.modules.xyz`. Feel free to discuss it here but I ddon't expect @kartikeyakirar to do this within this PR. That's why I created a separate issue (as opposed to a comment here) not to explode this PR.

m7pr replied:
The issue https://github.com/insightsengineering/teal.slice/issues/529

m7pr replied:
> You are mistaken. It is because shiny contains the core functionalities that teal builds upon.

agree, yeah but not for teal.slice

m7pr replied:
>This is exactly the difference between loading and attaching.

You can’t attach a package without first loading it - `library()` loads then attaches the package.

m7pr replied:
Let's remove `shiny::` and `library(shiny)` in examples, and let's bring them back if we decide to move `shiny` to `Imports`.

kartikeyakirar replied:
> Let's remove shiny:: and library(shiny) in examples,

this is already done in all the document examples. I have removed two cases from the vignettes.[0f073f3](https://github.com/insightsengineering/teal.slice/pull/506/commits/0f073f33db5ebd3a819cdb211c9436fcaa7d986b)

m7pr commented on R/FilterStateDate.R:
```diff
@@ -1,17 +1,25 @@
 #'
+#' @title `FilterState` object for Date variable
+#'
+#' @description Manages choosing a range of Dates.
```
@kartikeyakirar should `Date` in title and in description be included in backticks as this is a class in R?
Like you type `\code{character}` I think we should also type `\code{Date}` but use backticks instead of `\code{}` (I'm using \code{} only because backticks are reserved on GitHub markdown.

kartikeyakirar replied:
[14f1415](https://github.com/insightsengineering/teal.slice/pull/506/commits/14f14151f3e5160b99f83942eabeac347de53591)

m7pr commented on R/FilterStateDate.R:
```diff
@@ -1,17 +1,25 @@
+#' # use non-exported function from teal.slice
+#' include_js_files <- getFromNamespace("include_js_files", "teal.slice")
+#' include_css_files <- getFromNamespace("include_css_files", "teal.slice")
+#' include_js_files <- getFromNamespace("include_js_files", "teal.slice")
```
@kartikeyakirar why `include_js_files <- getFromNamespace("include_js_files", "teal.slice")` is repeated twice?

kartikeyakirar replied:
thanks for catching this. I have removed it [88507b0](https://github.com/insightsengineering/teal.slice/pull/506/commits/88507b0bd36dd8f52df0895e23a9b403c16c660d)

m7pr commented on R/FilterStateChoices.R:
```diff
@@ -1,16 +1,24 @@
+#' # use non-exported function from teal.slice
+#' include_js_files <- getFromNamespace("include_js_files", "teal.slice")
+#' include_css_files <- getFromNamespace("include_css_files", "teal.slice")
+#' include_js_files <- getFromNamespace("include_js_files", "teal.slice")
```
@kartikeyakirar why include_js_files <- getFromNamespace("include_js_files", "teal.slice") is repeated twice?

m7pr commented on R/FilterStateExpr.R:
```diff
@@ -106,7 +111,7 @@ FilterStateExpr <- R6::R6Class( # nolint
     #' Sets filtering state.
     #'
-    #' @param state a `teal_slice` object
+    #' @param state (`teal_slice`) object
```
@kartikeyakirar same as in other places for `param` - I would remove `object`

kartikeyakirar replied:
[95ced22](https://github.com/insightsengineering/teal.slice/pull/506/commits/95ced22df80e0414a42be5ca506cdcd43a3e9a92)

m7pr commented on R/FilterStateLogical.R:
```diff
@@ -1,35 +1,43 @@
+#' # use non-exported function from teal.slice
+#' include_js_files <- getFromNamespace("include_js_files", "teal.slice")
+#' include_css_files <- getFromNamespace("include_css_files", "teal.slice")
+#' include_js_files <- getFromNamespace("include_js_files", "teal.slice")
```
why include_js_files <- getFromNamespace("include_js_files", "teal.slice") is repeated twice? cc @kartikeyakirar

m7pr commented on R/FilterStateRange.R:
```diff
@@ -1,16 +1,24 @@
+#' # use non-exported function from teal.slice
+#' include_js_files <- getFromNamespace("include_js_files", "teal.slice")
+#' include_css_files <- getFromNamespace("include_css_files", "teal.slice")
+#' include_js_files <- getFromNamespace("include_js_files", "teal.slice")
```
why include_js_files <- getFromNamespace("include_js_files", "teal.slice") is repeated twice? cc @kartikeyakirar

m7pr commented on R/FilterStates-utils.R:
```diff
@@ -1,55 +1,59 @@
 #' @keywords internal
-#' @export
 #' @examples
-#' library(shiny)
```
@kartikeyakirar the same topic as discussed with @pawelru due to the transition of shiny from Depends to Imports
https://github.com/insightsengineering/teal.slice/issues/529
We then need to keep `library(shiny)`

pawelru replied:
at the current state the removal is correct - this line is redundant
We might reintroduce it in the aforementioned task. Before or after release.

m7pr replied:
ok

m7pr commented on R/FilterStates-utils.R:
```diff
@@ -132,19 +136,21 @@ init_filter_states.SummarizedExperiment <- function(data, # nolint
-#' @param data (`object`)\cr
-#'   the R object containing elements which class can be checked through `vapply` or `apply`.
-#'
+#' @param data (`object`)
```
```suggestion
#' @param data
```

m7pr commented on R/FilterStatesDF.R:
```diff
@@ -1,12 +1,19 @@
+#' # use non-exported function from teal.slice
+#' include_js_files <- getFromNamespace("include_js_files", "teal.slice")
+#' include_css_files <- getFromNamespace("include_css_files", "teal.slice")
+#' include_js_files <- getFromNamespace("include_js_files", "teal.slice")
```
include_js_files is here twice

m7pr commented:
Hey @kartikeyakirar I left few minor comments, but I would like them to be resolved before we merge

kartikeyakirar commented on R/FilterStateDatettime.R:
```diff
@@ -1,17 +1,25 @@
+#' # use non-exported function from teal.slice
+#' include_js_files <- getFromNamespace("include_js_files", "teal.slice")
+#' include_css_files <- getFromNamespace("include_css_files", "teal.slice")
+#' include_js_files <- getFromNamespace("include_js_files", "teal.slice")
```
[88507b0](https://github.com/insightsengineering/teal.slice/pull/506/commits/88507b0bd36dd8f52df0895e23a9b403c16c660d)

m7pr commented:
Hey, just to summarize what's going on during this PR.
@kartikeyakirar did a great job eliminating ~20 cleanup issues from the pre-release checklist.

Changes in repository included
- addition of new CRAN badges to README
- changing installation instructions to regular installation from CRAN
- addition of the CRAN RELEASE issue template
- cleanup of `pkgdown` titles

During this PR @chlebowa proposed we move internal examples to a special vignette (with an addition to documentation with `@seealso` tag pointing to this vignette). Later on we also discussed if we should move them to `inst/demo()` part of the package. At the end of the day, we decided to use `getFromNamespace()` function to overcome limitation of `:::` usage in the documentation, which is discouraged by CRAN policy. The vignette was discarded.

Other main changes included
- examples and vignettes were run and verified for compliteness
- refinement of roxygen2 `@title`, `@description` and `@details` sections/tags
- WORDLIST cleanup and typos verification
- review of cleanness of the documentation 
    - mostly classes of parameters
    - addition of `@internal` tags to some pages so they are not listed in the index of the package
    - quoting package names and `R` with backticks
- removal of `::` prefixing and `library(shiny)` calls from examples since `shiny` is in `Depends` and prefixing in examples clutters the picture
- **tests:** 
    - adding prefixes in `tests/` and substitution of `:::` usage with `getFromNamespace()`
    - we added a skip flow in case some packages are missing on a run (since they are only in Suggests)
        - `testthat::skip_if_not_installed("MultiAssayExperiment")`
        - @averissimo started a discussion about this in here https://github.com/insightsengineering/teal.slice/pull/506#issuecomment-1883300383 and provided a change for that in here https://github.com/insightsengineering/teal.slice/pull/506/commits/3220aac75b8cf7cd85a71cef273750b2d15bffb6
        - this will be required by IDR GHA flows anyway https://github.com/insightsengineering/idr-tasks/issues/713
- **vignettes:** 
    - filter panel for developer vignettes had it's graph of dependencies refreshed
    - classes were renamed (e.g. `DefaultFilteredDatase` to `DataframeFilteredDataset`, `FilteredDatasetMAE` to `MAEFilteredDataset`) and their appearance in the proper place of the vignettes was reviewed
    
    
 Final remarks
 - Last review changes required was raised by @chlebowa https://github.com/insightsengineering/teal.slice/pull/506#pullrequestreview-1808818075 and is now satisfied.
 - @pawelru found a NOTE in R CMD CHECK https://github.com/insightsengineering/teal.slice/pull/506#issuecomment-1886970072 that @kartikeyakirar claims to be resolved once we land `teal` on CRAN so I believe this is minor
https://github.com/insightsengineering/teal.slice/pull/506#issuecomment-1887057751
- I believe there is one more thing that is not yet covered https://github.com/insightsengineering/teal.slice/pull/506#discussion_r1462162016 but I created a new issue for that https://github.com/insightsengineering/teal.slice/issues/531

Multiple people were involved in this PR, kudos to @kartikeyakirar @chlebowa and @averissimo 
A lot of things have been provided in this PR (112 files changed!) and I think it's time to merge this. DONE is better than PERFECT. We can tackle other small things outside of this PR. It can not grow any longer and I state it's in good shape to be merged.

kartikeyakirar commented:
Thank you, @m7pr, for your excellent summary of the PR and approving it. I concur that it contains numerous changes, and as you suggested any additional change requests should be separate issue. Reviewing the entire pull request in one go could be challenging. I plan to merge this pull request in 3 hours, which should provide enough time for any final comments or feedback.

thanks everyone.
